### PR TITLE
feat: Add --provenance-bundle flag to npm publish

### PR DESCRIPTION
**Pull request by ianlewis**
This PR adds a new `--provenance-bundle` flag alongside the `--provenance` flag for `npm publish`. The flag takes a file path to a file in [sigstore bundle format](https://github.com/sigstore/protobuf-specs/blob/c8a23a43dc1c685d336d112864ba237a15061c53/protos/sigstore_bundle.proto#L63-L84) (JSON encoded).

This allows CI builders to generate signed provenance external to the npm CLI and upload it to the npm registry along with the created package.

### CODE REVIEW

File: workspaces/libnpmpublish/lib/publish.js [hunk 1]
```diff
@@ -137,6 +139,13 @@ const buildMetadata = async (registry, manifest, tarballData, spec, opts) => {
     length: tarballData.length,
   }
 
+  if (provenance === true && provenanceBundle) {
+    throw Object.assign(
+      new Error('--provenance and --provenance-bundle cannot be specified at the same time.'),
+      { code: 'EUSAGE' }
+    )
+  }
```
Comment (ljharb) [inline workspaces/libnpmpublish/lib/publish.js:147]: this seems confusing - if they're mutually exclusive, why not `--provenance=path/to/bundle`?

Comment (ianlewis) [reply workspaces/libnpmpublish/lib/publish.js:147]: I did try this. However, I wanted to preserve the existing `--provenance` boolean, and npm seems finicky about its flags. If I create a flag that can be a boolean or string then `npm publish --provenance package-name` will  no longer work and parse `package-name` as the argument to `--provenance`.

**Comment by wraithgar** [comment]
This is not the direction we are going w/ this flag.  We need to fix the bug where we can't configure a flag to be either a boolean or a string, and use that flag for this purpose.  The fix is in https://github.com/npm/cli/pull/6020 (I think? Someone used copilot to "fix" the PR and now it's pretty unreadable).

**Comment by wraithgar** [comment]
The original provenance PR did have this functionality and that is how we found the config bug, so it was paused till we can fix it.

**Comment by ianlewis** [comment]
> The original provenance PR did have this functionality and that is how we found the config bug, so it was paused till we can fix it.

Just to be clear, Is it fair to say we still want to add the functionality once the config bug is fixed and we just don't want it to be a separate flag?

**Comment by wraithgar** [comment]
> Just to be clear, Is it fair to say we still want to add the functionality once the config bug is fixed and we just don't want it to be a separate flag?

Yes the end goal is for you to be able to go `--provenance` or `--provenance=/tmp/build/foo-1.2.0.output`

**Comment by ianlewis** [comment]
> > Just to be clear, Is it fair to say we still want to add the functionality once the config bug is fixed and we just don't want it to be a separate flag?
> 
> Yes the end goal is for you to be able to go `--provenance` or `--provenance=/tmp/build/foo-1.2.0.output`

I created #6313 to explain the flag parsing issue. This is unfortunately blocking me from adding publishing support to https://github.com/slsa-framework/slsa-github-generator.

I think this will also require changes to `nopt` since flags are not parsed correctly if they could be a boolean or something else like a string. i.e. `npm publish --provenance --help` would not work as `--help` would get parsed an argument to `--provenance` rather than handled as a flag itself`. I think most flag parsing libraries don't allow this so they don't have to handle cases like this.

**Comment by bdehamer** [comment]
@ianlewis just an FYI that this work is continuing [here](https://github.com/npm/cli/pull/6490).

**Pull request closed without merging**